src/weak_memory.rs

```diff
@@ -0,0 +1,271 @@
+
+                true
+            })
+            .filter(|&store_elem| {
+                if is_seqcst {
+                    // An SC load needs to ignore all but last store maked SC (stores not marked SC are not
+                    // affected)
+                    let include = !(store_elem.is_seqcst && found_sc);
+                    found_sc |= store_elem.is_seqcst;
+                    include
+                } else {
+                    true
+                }
+            });
+
+        candidates.choose(rng)
+    }
+
+    /// ATOMIC STORE IMPL in the paper
+    /// The paper also wants the variable's clock (AtomicMemoryCellClocks::sync_vector in our code)
+    /// to be in the store element, but it's not used anywhere so we don't actually need it
+    fn store_impl(
+        &mut self,
+        val: ScalarMaybeUninit<Tag>,
+        index: VectorIdx,
+        thread_clock: &VClock,
+        is_seqcst: bool,
+    ) {
+        let store_elem = StoreElement {
+            store_index: index,
+            timestamp: thread_clock[index],
+            // In the language provided in the paper, an atomic store takes the value from a
+            // non-atomic memory location.
+            // But we already have the immediate value here so we don't need to do the memory
+            // access
+            val,
+            is_seqcst,
+            loads: RefCell::new(FxHashMap::default()),
+        };
+        self.buffer.push_back(store_elem);
+        if self.buffer.len() > STORE_BUFFER_LIMIT {
+            self.buffer.pop_front();
+        }
+        if is_seqcst {
+            // Every store that happens before this needs to be marked as SC
+            // so that in fetch_store, only the last SC store (i.e. this one) or stores that
+            // aren't ordered by hb with the last SC is picked.
+            //
+            // This is described in the paper (§4.5) and implemented in tsan11
+            // (https://github.com/ChrisLidbury/tsan11/blob/ecbd6b81e9b9454e01cba78eb9d88684168132c7/lib/tsan/rtl/tsan_relaxed.cc#L160-L167)
+            // but absent in the operational semantics.
+            self.buffer.iter_mut().rev().for_each(|elem| {
+                if elem.timestamp <= thread_clock[elem.store_index] {
+                    elem.is_seqcst = true;
+                }
+            })
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct StoreElement {
+    /// The identifier of the vector index, corresponding to a thread
+    /// that performed the store.
+    store_index: VectorIdx,
+
+    /// Whether this store is SC.
+    is_seqcst: bool,
+
+    /// The timestamp of the storing thread when it performed the store
+    timestamp: VTimestamp,
+    /// The value of this store
+    val: ScalarMaybeUninit<Tag>,
+
+    /// Latest timestamp of each thread that has loaded from this store element
+    /// Behind a RefCell to keep load op take &self
+    loads: RefCell<FxHashMap<VectorIdx, VTimestamp>>,
```

Comment:
Wait, a store element stores the loads that loaded from it?!?? What is *extremely* surprising to me. I didn't read the paper you are implementing, but I have seen operational semantics for weak memory models and they don't do anything like this.

I guess the paper has it so :shrug: but this goes to show that I don't have any intuition for why the algorithm in this paper makes sense.

Reply:
It's not too hard to see why this is needed

> Coherence of Read-Reads (CoRR). This states that if two reads from the same location are ordered by `hb`, the reads cannot observe writes ordered differently in `mo`

So when going through the store buffer in reverse modification order, we need to check if a store element has been read by something that happens before us. Happens before are determined by comparing the vector clocks of two threads, so we need to keep track of the VCs that loaded from a store.

Reply:
But why does Strong Release Acquire or Promising not need anything like that? Or do I just misremember and they have something like that?

Reply:
Each thread in SRA's machine maintains an explicit data structure of its local view of all memory locations (history of view is not kept). On top of read, write and RMWs, there is another action called _process_ which updates a thread's local view of memory using from writes in message buffers. 

It is true that there is no bookkeeping on when a SRA read happens, but it's the process action which makes sure that the current value in local memory (observable by actual reads at any moment) doesn't get overwritten by a value from an earlier write. So each thread-local location contains both a value and the global timestamp of the write, these are modified when process happens.

Promising is very similar (§2.1): each thread has a view of memory containing a mapping from memory locations to the timestamp of the most recent write it has observed, and all writes have a globally unique timestamp.

We don't have a sporadic process action, so reads have to do the bookkeeping themselves. There is no globally unique timestamp for store elements. Our threads also don't have isolated view of all memory locations - the store buffer of a location is shared across all threads, so timestamps must be recorded on reads to determine if one read happens-before another one, and we have to keep all previous writes when a new one comes in because another thread may need them.

We could have a table on the store buffer (instead of store elements) that records "thread t has read from the nth-last element in the buffer at timestamp x" which could achieve the same effect as what we are doing now (a table that records "thread t has read from _this_ store element at timestamp x" on each store element). Though then we need to query this table every time before crawling through the buffer, including when the nth-last limit we got is irrelevant because we need to stop before reaching nth-last.

Reply:
> It is true that there is no bookkeeping when a SRA read happens,

That's not what I mean though -- what I mean is that in the SRA machine, a view stores all the events it saw, but there is nothing in "the other direction" -- an event (a store) never "points to" things that read from it.

SRA without a sporadic action can just do that action each time a location is read, without changing observable behavior.

> There is no globally unique timestamp for store elements. Our threads also don't have isolated view of all memory locations 

That seems to be a more fundamental difference.

> We could have a table on the store buffer (instead of store elements) that records "thread t has read from the nth-last element in the buffer at timestamp x" which could achieve the same effect as what we are doing now (a table that records "thread t has read from this store element at timestamp x" on each store element). Though then we need to query this table every time before crawling through the buffer, including when the nth-last limit we got is irrelevant because we need to stop before reaching nth-last.

So what I am reading here is, a representation more like SRA would be possible, but it'd be less efficient? That seems fair to me. I would find it more understandable but that doesn't mean we should do it, given in particular that you are following a well-documented algorithm that has already been implemented by other tools.

Thanks for your detailed reply!

Reply:
> So what I am reading here is, a representation more like SRA would be possible, but it'd be less efficient? 

It's mostly an alternative representation to what we have now. I guess it is slightly more similar to SRA but as you pointed out, the fundamental difference is that our threads don't maintain their isolated views, and indeed that'd be a lot more expensive.